File: ⭐️ Favorites ⭐️/ArrayManipulation/⭐️ groupAnagrams/✅ groupAnagrams.js
// BigO
// w = number of words / length of array
// n = length of longest word
// Space: O(wn) => 
// Time: O(wn log(n)) => need to loop through and sort each word => sorting === O(n log(n)) & there are w words

function groupAnagrams(array) {
    // Edge Case
    if (array.length === 0 ) { return -1 }
    // Variables
    result = {}

    // Loigic
    for (const [i, str] of array.entries()) {
    console.log(i, str)
    // -- //
        // Sort current string    
        sorted = str.split('').sort().join('')
        // Is sorted word in hash table
        if (sorted in result) {
            // Add orrigional string to array of strings at sorted key
            result[sorted].push(str)
        } else {
            // Add sorted key with array of orrigional strings
            result[sorted] = [str]
        }
    }
return result}

// -- *** TESTS *** -- //
// -- *** TESTS *** -- //
// -- *** TESTS *** -- //
testArray_1 = ["eat", "tea", "tan", "ate", "nat", "bat"]
testArray_2 = ['yo', 'act', 'flop', 'tac', 'foo', 'cat', 'oy', 'olfp']
console.log(groupAnagrams(testArray_1))
console.log(groupAnagrams(testArray_2))

File: ⭐️ Favorites ⭐️/ArrayManipulation/⭐️ groupAnagrams/✅ groupAnagrams.py
def groupAnagrams(strs):
    result = {}
    for i in strs: 

        # Turn string into sorted string
        x = ''.join(sorted(i))
        print(x)

        if x in result: 
            # Add ORIGIONAL STRING To That Result Group 
            result[x].append(i)

            print(result)
        else:
            # Add ORRIGIONAL STRING To New Result Group
            result[x] = [i]
            print(result)

    # Return
    ## Want to return a SINGLE LIST of all the indvidual VALUES 
    ## (group of anagrams) instead of current Result Object. 
    return list(result.values())

print(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))